File: ureka-framework-main/ureka_framework/resource/crypto/ecdh.py
# Random
import os
from typing import Tuple

# Hash
from cryptography.hazmat.primitives import hashes

# ECDH
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes

# AES
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding

# Serialization
from ureka_framework.resource.crypto.serialization_util import (
    byte_to_base64str,
    str_to_byte,
)

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Resource (Measurer)
from ureka_framework.resource.logger.simple_measurer import measure_resource_func


######################################################
# Random Number Generation
#   pyca/cryptography recommends using operating system’s provided random number generator
######################################################
@measure_resource_func
def generate_random_byte(bytes_num: int) -> bytes:
    return os.urandom(bytes_num)


@measure_resource_func
def generate_random_str(bytes_num: int) -> str:
    return byte_to_base64str(os.urandom(bytes_num))


######################################################
# Hash Function
######################################################
@measure_resource_func
def generate_sha256_hash_bytes(message: bytes) -> bytes:
    digest = hashes.Hash(hashes.SHA256())
    digest.update(message)
    generated_hash: bytes = digest.finalize()

    return generated_hash


@measure_resource_func
def generate_sha256_hash_str(message_str: str) -> str:
    message_bytes: bytes = str_to_byte(message_str)
    generated_hash_bytes: bytes = generate_sha256_hash_bytes(message_bytes)
    generated_hash_str: str = byte_to_base64str(generated_hash_bytes)

    return generated_hash_str


######################################################
# ECDH Key Factory
#   In a key agreement protocol deriving cryptographic keys from a Diffie-Hellman exchange
#       can derive a salt value from public nonces exchanged and authenticated between communicating parties
#       as part of the key agreement.
#       <Ref> https://datatracker.ietf.org/doc/html/rfc5869.html#section-3.1
#   Further recommand ECDHE (ECDH, ephemeral) if consider Forward Secrecy
#       -> But need exchange ephemeral public keys & bring more overhead.
#       -> A practical solution for ephemeral maybe periodically re-exchange a HKDF-derived session key.
#       -> How to set the period depends on the trade-off between security and overhead.
######################################################
@measure_resource_func
def generate_ecdh_key(
    server_private_key: ec.EllipticCurvePrivateKey,
    salt: bytes,
    info: bytes,
    peer_public_key: ec.EllipticCurvePublicKey,
) -> bytes:
    # Perform ECDH
    shared_key = server_private_key.exchange(ec.ECDH(), peer_public_key)

    # Perform Key Derivation [HKDF, HMAC-based Extract-and-Expand KDF]
    derived_key = HKDF(
        algorithm=hashes.SHA256(),
        length=32,  # The desired length of the derived key in bytes.
        salt=salt,  # Randomizes the KDF’s output.
        info=info,  # Optional: application specific context information.
        backend=default_backend(),
    ).derive(shared_key)

    return derived_key


######################################################
# ECDH AES Encryption (CBC Mode)
#   Need Padding:
#       https://cryptography.io/en/3.4.2/hazmat/primitives/padding.html
#   Need HMAC:
#       If ureka protocol applies Command UTicket/Data UTicket with signatures on them,
#           the AES-CBC can be used and the HMAC can be omitted.
#       However, AES+HMAC may provide better performance than Command UTicket/Data UTicket.
#                   (Because HMAC verfication may be cheaper than Signature verification.)
#           In this case, AES-GCM (Galois Counter Mode) is better,
#               which is a  AEAD (authenticated encryption with additional data) mode so that HMAC are not necessary.
######################################################
@measure_resource_func
def cbc_encrypt(plaintext: bytes, key: bytes) -> Tuple[bytes, bytes]:
    # Randomly generate IV (Initialization Vector)
    #   IV must be the same number of bytes as the block_size of the cipher.
    #   IV must be random bytes.
    #   IV do not need to be kept secret and they can be included in a transmitted message.
    #   For better security, each time something is encrypted a new IV should be generated.
    #   Do not reuse an IV with a given key, and particularly do not use a constant IV, i.e.:
    #       iv = b"1234567890abcdef"  # Dangerous.
    #       iv = generate_random_byte(16) # Recommanded, but need be transmited & perform authenticity check everytime.
    iv = generate_random_byte(16)

    # Initailize AES (CBC mode)
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    encryptor = cipher.encryptor()

    # Padding using PKCS7 (CBC mode needs padding)
    padder = padding.PKCS7(128).padder()
    padded_message = padder.update(plaintext) + padder.finalize()

    # Encrypt message
    encrypted_message = encryptor.update(padded_message) + encryptor.finalize()

    return (encrypted_message, iv)


######################################################
# ECDH AES Decryption (CBC Mode)
######################################################
@measure_resource_func
def cbc_decrypt(ciphertext: bytes, key: bytes, iv: bytes) -> bytes:
    # Initailize AES (CBC mode)
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    decryptor = cipher.decryptor()

    # Decrypt message
    decrypted_message = decryptor.update(ciphertext) + decryptor.finalize()

    # Padding using PKCS7 (CBC mode needs padding)
    unpadder = padding.PKCS7(128).unpadder()
    unpadded_message = unpadder.update(decrypted_message) + unpadder.finalize()

    return unpadded_message


######################################################
# ECDH AES Encryption (GCM Mode)
#   Need NOT Padding & Need NOT HMAC:
#       But when decryption, it not only needs (ciphertext, key, iv), but also needs (associated_plaintext, authentication_tag)
#       <Ref> https://cryptography.io/en/3.4.2/hazmat/primitives/symmetric-encryption.html#cryptography.hazmat.primitives.ciphers.modes.GCM
#   Can add Associated Plaintext (authenticated but not encrypted) in message:
#       In ureka protocol, we assume all command & data are authenticated & encrypted,
#           so associated_plaintext can be set as None.
######################################################
@measure_resource_func
def gcm_gen_iv() -> bytes:
    # Generate a random 96-bit IV.
    iv = generate_random_byte(12)
    return iv


@measure_resource_func
def gcm_encrypt(
    plaintext: bytes, associated_plaintext: bytes, key: bytes, iv: bytes = None
) -> Tuple[bytes, bytes, bytes]:
    # Construct an AES-GCM Cipher object with the given key and a
    # randomly generated IV.
    encryptor = Cipher(
        algorithms.AES(key), modes.GCM(iv), backend=default_backend()
    ).encryptor()

    # associated_plaintext will be authenticated but not encrypted,
    # it must also be passed in on decryption.
    encryptor.authenticate_additional_data(associated_plaintext)

    # Encrypt the plaintext and get the associated ciphertext.
    # GCM does not require padding.
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()

    return (ciphertext, encryptor.tag)


######################################################
# ECDH AES Decryption (GCM Mode)
######################################################
@measure_resource_func
def gcm_decrypt(
    ciphertext: bytes, associated_data: bytes, tag: bytes, key: bytes, iv: bytes
) -> bytes:
    # Construct a Cipher object, with the key, iv, and additionally the
    # GCM tag used for authenticating the message.
    decryptor = Cipher(
        algorithms.AES(key), modes.GCM(iv, tag), backend=default_backend()
    ).decryptor()

    # We put associated_data back in or the tag will fail to verify
    # when we finalize the decryptor.
    decryptor.authenticate_additional_data(associated_data)

    # Decryption gets us the authenticated plaintext.
    # If the tag does not match an InvalidTag exception will be raised.
    return decryptor.update(ciphertext) + decryptor.finalize()

File: UrekaApp/app/src/main/java/ureka/framework/resource/crypto/ECDH.java
package ureka.framework.resource.crypto;

import org.bouncycastle.crypto.digests.SHA256Digest;
import org.bouncycastle.crypto.generators.HKDFBytesGenerator;
import org.bouncycastle.crypto.params.HKDFParameters;
import org.bouncycastle.jce.provider.BouncyCastleProvider;

import java.nio.charset.StandardCharsets;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.SecureRandom;
import java.security.Security;
import java.security.interfaces.ECPrivateKey;
import java.security.interfaces.ECPublicKey;
import java.util.Arrays;
import java.util.Base64;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.KeyAgreement;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import ureka.framework.resource.logger.SimpleLogger;
import ureka.framework.resource.logger.SimpleMeasurer;

public class ECDH {
    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    //////////////////////////////////////////////////////
    // Random Number Generation
    // pyca/cryptography recommends using operating system's provided random number generator
    //////////////////////////////////////////////////////
    public static byte[] generateRandomByte(int bytesNum) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_generateRandomByte, bytesNum);
    }
    private static byte[] _generateRandomByte(int bytesNum) {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[bytesNum];
        random.nextBytes(bytes);
        return bytes;
    }

    public static String generateRandomStr(int bytesNum) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_generateRandomStr, bytesNum);
    }
    private static String _generateRandomStr(int bytesNum) {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[bytesNum];
        random.nextBytes(bytes);
        return SerializationUtil.byteToBase64Str(bytes);
    }

    //////////////////////////////////////////////////////
    // Hash Function
    //////////////////////////////////////////////////////
    public static byte[] generateSha256HashBytes(byte[] message) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_generateSha256HashBytes, message);
    }
    private static byte[] _generateSha256HashBytes(byte[] message)
            throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256", "SC");
        return digest.digest(message);
    }

    public static String generateSha256HashStr(String messageStr) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_generateSha256HashStr, messageStr);
    }
    private static String _generateSha256HashStr(String messageStr)
            throws Exception {
        byte[] messageBytes = SerializationUtil.strToByte(messageStr);
        byte[] hashBytes = _generateSha256HashBytes(messageBytes);
        return SerializationUtil.byteToBase64Str(hashBytes);
    }

    //////////////////////////////////////////////////////
    // ECDH Key Factory
    //   In a key agreement protocol deriving cryptographic keys from a Diffie-Hellman exchange
    //       can derive a salt value from public nonces exchanged and authenticated between communicating parties
    //       as part of the key agreement.
    //       <Ref> https://datatracker.ietf.org/doc/html/rfc5869.html#section-3.1
    //   Further recommend ECDHE (ECDH, ephemeral) if consider Forward Secrecy
    //       -> But need exchange ephemeral public keys & bring more overhead.
    //       -> A practical solution for ephemeral maybe periodically re-exchange a HKDF-derived session key.
    //       -> How to set the period depends on the trade-off between security and overhead.
    //////////////////////////////////////////////////////
    public static byte[] generateEcdhKey(ECPrivateKey serverPrivateKey, byte[] salt, byte[] info, ECPublicKey peerPublicKey) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_generateEcdhKey, serverPrivateKey, salt, info, peerPublicKey);
    }
    private static byte[] _generateEcdhKey(ECPrivateKey serverPrivateKey, byte[] salt, byte[] info, ECPublicKey peerPublicKey)
        throws NoSuchAlgorithmException, NoSuchProviderException, InvalidKeyException {
        try {
            // Perform ECDH
            KeyAgreement keyAgreement = KeyAgreement.getInstance("ECDH", "SC");
            keyAgreement.init(serverPrivateKey);
            keyAgreement.doPhase(peerPublicKey, true);
            byte[] sharedSecret = keyAgreement.generateSecret();

            // Perform Key Derivation [HKDF, HMAC-based Extract-and-Expand KDF]
            HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA256Digest());
            HKDFParameters hkdfParams = new HKDFParameters(sharedSecret, salt, info);
            hkdf.init(hkdfParams);
            byte[] derivedKey = new byte[32]; // Length of the derived key in bytes
            hkdf.generateBytes(derivedKey, 0, derivedKey.length);

            return derivedKey;
        } catch (NoSuchAlgorithmException e) {
            String failureMsg = "generateEcdhKey: NoSuchAlgorithmException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchProviderException e) {
            String failureMsg = "generateEcdhKey: NoSuchProviderException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidKeyException e) {
            String failureMsg = "generateEcdhKey: InvalidKeyException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        }

    }

    //////////////////////////////////////////////////////
    // ECDH AES Encryption (CBC Mode)
    //   Need Padding:
    //       https://cryptography.io/en/3.4.2/hazmat/primitives/padding.html
    //   Need HMAC:
    //       If ureka protocol applies Command UTicket/Data UTicket with signatures on them,
    //           the AES-CBC can be used and the HMAC can be omitted.
    //       However, AES+HMAC may provide better performance than Command UTicket/Data UTicket.
    //                   (Because HMAC verification may be cheaper than Signature verification.)
    //           In this case, AES-GCM (Galois Counter Mode) is better,
    //               which is a  AEAD (authenticated encryption with additional data) mode so that HMAC are not necessary.
    //////////////////////////////////////////////////////
    public static byte[][] cbcEncrypt(byte[] plaintext, byte[] key) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_cbcEncrypt, plaintext, key);
    }
    private static byte[][] _cbcEncrypt(byte[] plaintext, byte[] key)
        throws Exception {
        // Randomly generate IV (Initialization Vector)
        //   IV must be the same number of bytes as the blockSize of the cipher.
        //   IV must be random bytes.
        //   IV do not need to be kept secret and they can be included in a transmitted message.
        //   For better security, each time something is encrypted a new IV should be generated.
        //   Do not reuse an IV with a given key, and particularly do not use a constant IV, i.e.:
        //       iv = b"1234567890abcdef"  # Dangerous.
        //       iv = generateRandomByte(16) # Recommended, but need be transmitted & perform authenticity check every time.
        try {
            byte[] iv = generateRandomByte(16);

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS7Padding", "SC");
            SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);

            cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);

            byte[] paddedMessage = cipher.doFinal(plaintext);

            return new byte[][] { paddedMessage, iv };
        } catch (IllegalBlockSizeException e) {
            String failureMsg = "cbcEncrypt: IllegalBlockSizeException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidAlgorithmParameterException e) {
            String failureMsg = "cbcEncrypt: InvalidAlgorithmParameterException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidKeyException e) {
            String failureMsg = "cbcEncrypt: InvalidKeyException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (BadPaddingException e) {
            String failureMsg = "cbcEncrypt: BadPaddingException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchAlgorithmException e) {
            String failureMsg = "cbcEncrypt: NoSuchAlgorithmException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchPaddingException e) {
            String failureMsg = "cbcEncrypt: NoSuchPaddingException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchProviderException e) {
            String failureMsg = "cbcEncrypt: NoSuchProviderException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        }
    }

    //////////////////////////////////////////////////////
    // ECDH AES Decryption (CBC Mode)
    //////////////////////////////////////////////////////
    public static byte[] cbcDecrypt(byte[] ciphertext, byte[] key, byte[] iv) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_cbcDecrypt, ciphertext, key, iv);
    }
    private static byte[] _cbcDecrypt(byte[] ciphertext, byte[] key, byte[] iv)
        throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException
        , InvalidAlgorithmParameterException, InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS7Padding", "SC");
            SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);

            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);

            return cipher.doFinal(ciphertext);
        } catch (IllegalBlockSizeException e) {
            String failureMsg = "cbcDecrypt: IllegalBlockSizeException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidAlgorithmParameterException e) {
            String failureMsg = "cbcDecrypt: InvalidAlgorithmParameterException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidKeyException e) {
            String failureMsg = "cbcDecrypt: InvalidKeyException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (BadPaddingException e) {
            String failureMsg = "cbcDecrypt: BadPaddingException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchAlgorithmException e) {
            String failureMsg = "cbcDecrypt: NoSuchAlgorithmException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchPaddingException e) {
            String failureMsg = "cbcDecrypt: NoSuchPaddingException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchProviderException e) {
            String failureMsg = "cbcDecrypt: NoSuchProviderException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        }
    }

    //////////////////////////////////////////////////////
    // ECDH AES Encryption (GCM Mode)
    //   Need NOT Padding & Need NOT HMAC:
    //       But when decryption, it not only needs (ciphertext, key, iv), but also needs (associatedPlaintext, authenticationTag)
    //       <Ref> https://cryptography.io/en/3.4.2/hazmat/primitives/symmetric-encryption.html#cryptography.hazmat.primitives.ciphers.modes.GCM
    //   Can add Associated Plaintext (authenticated but not encrypted) in message:
    //       In ureka protocol, we assume all command & data are authenticated & encrypted,
    //           so associatedPlaintext can be set as None.
    //////////////////////////////////////////////////////
    public static byte[] gcmGenIv() throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_gcmGenIv);
    }
    private static byte[] _gcmGenIv() {
        // Generate a random 96-bit IV.
        return _generateRandomByte(12);
    }

    public static byte[][] gcmEncrypt(byte[] plaintext, byte[] associatePlaintext, byte[] key, byte[] iv) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_gcmEncrypt, plaintext, associatePlaintext, key, iv);
    }
    private static byte[][] _gcmEncrypt(byte[] plaintext, byte[] associatedPlaintext, byte[] key, byte[] iv)
        throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException
        , InvalidAlgorithmParameterException, InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
        try {
            if (iv == null) {
                iv = _generateRandomByte(12); // GCM standard recommends a 12-byte IV
            }
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding", "SC");
            SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
            GCMParameterSpec gcmSpec = new GCMParameterSpec(128, iv); // 128-bit authentication tag length

            cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec);
            cipher.updateAAD(associatedPlaintext);

            byte[] ciphertext = cipher.doFinal(plaintext);
            byte[] tag = new byte[16];
            System.arraycopy(ciphertext, ciphertext.length - 16, tag, 0, 16);
            byte[] finalCiphertext = new byte[ciphertext.length - 16];
            System.arraycopy(ciphertext, 0, finalCiphertext, 0, ciphertext.length - 16);

            return new byte[][] { finalCiphertext, iv, tag };
        } catch (Exception e) {
            SimpleLogger.simpleLog("error", e.getMessage());
            throw e;
        }
    }

    public static byte[] gcmDecrypt(byte[] ciphertext, byte[] associatedData, byte[] tag, byte[] key, byte[] iv) throws Exception {
        return SimpleMeasurer.measureResourceFunc(ECDH::_gcmDecrypt, ciphertext, associatedData, tag, key, iv);
    }
    private static byte[] _gcmDecrypt(byte[] ciphertext, byte[] associatedData, byte[] tag, byte[] key, byte[] iv)
        throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException
        , InvalidAlgorithmParameterException, InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
        try {
            byte[] combinedCiphertext = new byte[ciphertext.length + tag.length];
            System.arraycopy(ciphertext, 0, combinedCiphertext, 0, ciphertext.length);
            System.arraycopy(tag, 0, combinedCiphertext, ciphertext.length, tag.length);

            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding", "SC");
            SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
            GCMParameterSpec gcmSpec = new GCMParameterSpec(128, iv); // 128-bit authentication tag length

            cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec);
            cipher.updateAAD(associatedData);

            return cipher.doFinal(combinedCiphertext);
        } catch (IllegalBlockSizeException e) {
            String failureMsg = "gcmDecrypt: IllegalBlockSizeException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidAlgorithmParameterException e) {
            String failureMsg = "gcmDecrypt: InvalidAlgorithmParameterException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (InvalidKeyException e) {
            String failureMsg = "gcmDecrypt: InvalidKeyException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (BadPaddingException e) {
            String failureMsg = "gcmDecrypt: BadPaddingException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchAlgorithmException e) {
            String failureMsg = "gcmDecrypt: NoSuchAlgorithmException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchPaddingException e) {
            String failureMsg = "gcmDecrypt: NoSuchPaddingException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        } catch (NoSuchProviderException e) {
            String failureMsg = "gcmDecrypt: NoSuchProviderException occurs.";
            // SimpleLogger.simpleLog("error", "{" + failureMsg + "}: {" + e + "}");
            throw e;
        }
    }
}
